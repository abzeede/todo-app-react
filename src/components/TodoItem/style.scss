.todo-item {
  background-color: #fff;
  padding: 0.5em;
  display: flex;
  align-items: center;
  margin: 5px 0;
  border-radius: 4px;
  height: 30px;
  box-shadow: 0 1px 3px rgba(0, 0, 0, 0.12), 0 1px 2px rgba(0, 0, 0, 0.24);

  &:hover {
    background-color: rgb(217, 220, 225);

    .todo-item__remove-btn {
      visibility: visible;
      background-color: rgb(217, 220, 225);
    }

    .todo-item__input {
      background-color: rgb(217, 220, 225);
      border-bottom: 1px solid #8e8e8e;

      &--done, &--done:focus, &--done:hover {
        border-bottom: none;
      }
    }
  }

  &__checkbox {
    display: flex;
    flex-basis: 25px;
    cursor: pointer;
    margin: 0 5px;
  }

  &__input {
    display: flex;
    flex: 1;
    margin-right: 10px;
    border: none;
    font-size: 1em;
    color: #262626;

    &:focus {
      outline: none;
      border-bottom: 1px solid #8e8e8e;
    }

    &--done {
      text-decoration: line-through;
      color: #8e8e8e;
    }
  }

  &__remove-btn {
    height: 30px;
    width: 35px;
    border-radius: 4px;
    border: 1px solid #e8e8e8;
    font-size: 1em;
    color: #262626;
    cursor: pointer;
    visibility: hidden;

    &:hover {
      background-color: #ef5350 !important;
      border-color: #ef5350;
      color: white;
    }
  }
}
