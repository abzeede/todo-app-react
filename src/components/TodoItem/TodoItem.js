import React from 'react'
import { number, string, bool, func } from 'prop-types'
import Checkbox from 'components/Checkbox'
import './style.scss'

const TodoItem = ({ id, detail, done, onDelete, onToggleDone, onEdit }) => {
  const onChange = e => {
    onEdit(id, e.target.value)
  }

  return (
    <li className="todo-item">
      <span className="todo-item__checkbox">
        <Checkbox onClick={() => onToggleDone(id)} checked={done} />
      </span>
      <input
        className={`todo-item__input ${done ? 'todo-item__input--done' : ''}`}
        onChange={onChange}
        value={detail}
      />
      <button className="todo-item__remove-btn" onClick={() => onDelete(id)}>
        <i className="fa fa-trash" />
      </button>
    </li>
  )
}

TodoItem.propTypes = {
  id: number,
  detail: string,
  done: bool,
  onToggleDone: func,
  onDelete: func,
  onEdit: func,
}

export default TodoItem
