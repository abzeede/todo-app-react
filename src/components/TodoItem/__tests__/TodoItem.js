import React from 'react'
import { mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import TodoItem from 'components/TodoItem/TodoItem'

describe('TodoItem', () => {
  it('should render properly', () => {
    const doneItem = mount(<TodoItem id={1} detail="something" done />)
    const todoItem = mount(<TodoItem id={2} detail="something" done={false} />)

    expect(todoItem.find('input').props().value).toEqual('something')
    expect(toJson(doneItem)).toMatchSnapshot()
    expect(toJson(todoItem)).toMatchSnapshot()
  })

  it('should call onToggleDone when click on CheckBox', () => {
    const onToggleDone = jest.fn(value => value)
    const todoItem = mount(<TodoItem id={2} detail="something" done={false} onToggleDone={onToggleDone} />)
    todoItem.find('.checkbox').simulate('click')
    expect(onToggleDone).toHaveBeenCalledWith(2)
  })

  it('should call onToggleDone when click on CheckBox', () => {
    const onToggleDone = jest.fn(value => value)
    const todoItem = mount(<TodoItem id={2} detail="something" done={false} onToggleDone={onToggleDone} />)
    todoItem.find('.checkbox').simulate('click')
    expect(onToggleDone).toHaveBeenCalledWith(2)
  })

  it('should call onEdit when change input value', () => {
    const onEdit = jest.fn((id, value) => ({ id, value }))
    const wrapper = mount(<TodoItem id={2} detail="something" done={false} onEdit={onEdit} />)
    const input = wrapper.find('input')

    input.simulate('change', { target: { value: `I don't know what to do` } })
    expect(onEdit).toHaveBeenCalledWith(2, `I don't know what to do`)
  })

  it('should call onDelete when click Delete button', () => {
    const onDelete = jest.fn(id => id)
    const wrapper = mount(<TodoItem id={2} detail="something" done={false} onDelete={onDelete} />)
    wrapper.find('.todo-item__remove-btn').simulate('click')
    expect(onDelete).toHaveBeenCalledWith(2)
  })
})
