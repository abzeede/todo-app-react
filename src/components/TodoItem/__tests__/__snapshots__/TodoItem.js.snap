// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem should render properly 1`] = `
<TodoItem
  detail="something"
  done={true}
  id={1}
>
  <li
    className="todo-item"
  >
    <span
      className="todo-item__checkbox"
    >
      <Checkbox
        checked={true}
        onClick={[Function]}
      >
        <span
          className="checkbox checkbox--checked"
          onClick={[Function]}
        />
      </Checkbox>
    </span>
    <input
      className="todo-item__input todo-item__input--done"
      onChange={[Function]}
      value="something"
    />
    <button
      className="todo-item__remove-btn"
      onClick={[Function]}
    >
      <i
        className="fa fa-trash"
      />
    </button>
  </li>
</TodoItem>
`;

exports[`TodoItem should render properly 2`] = `
<TodoItem
  detail="something"
  done={false}
  id={2}
>
  <li
    className="todo-item"
  >
    <span
      className="todo-item__checkbox"
    >
      <Checkbox
        checked={false}
        onClick={[Function]}
      >
        <span
          className="checkbox false"
          onClick={[Function]}
        />
      </Checkbox>
    </span>
    <input
      className="todo-item__input "
      onChange={[Function]}
      value="something"
    />
    <button
      className="todo-item__remove-btn"
      onClick={[Function]}
    >
      <i
        className="fa fa-trash"
      />
    </button>
  </li>
</TodoItem>
`;
